File: course/tree/queue.py
class Node:
  def __init__(self, data):
    self.data = data
    self.next = None

class Queue:
  def __init__(self):
    self.head = None
    self.tail = None
    self._size = 0
    

  def push(self, elem):
    node = Node(elem)
    if self.tail is None:
      self.tail = node
    else:
      self.tail.next = node
      self.tail = node
    
    if self.head is None:
      self.head = node

    self._size += 1

  def pop(self):
    if self._size > 0:
      elem = self.head.data
      self.head = self.head.next
      self._size -= 1
      return elem    
    raise IndexError("Empty queue")
  
  def peek(self):
    if self._size > 0:
      elem = self.head.data
      return elem
    raise IndexError("Empty queue")

  def __len__(self):
    return self._size

  def __repr__(self):
    if self._size > 0:
      r = ""
      pointer = self.head
      while pointer:
        r += str(pointer.data) + " "
        pointer = pointer.next
      return r
    else:
      return 'Empty queue'

  def __str__(self):
    return self.__repr__()

File: queue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int data;
  struct Node* next;
} Node;

typedef struct Queue {
  struct Node* start;
  struct Node* end;
  int size;
} Queue;

void print(Queue* queue) {
  Node* node = queue->start; 
  while(node != NULL) {
    printf("%d ", node->data);
    node = node->next;
  }
  printf("\n");
}

Queue* constructor() {
  Queue* queue = malloc(sizeof(Queue));
  queue->start = NULL;
  queue->end = NULL;
  queue->size = 0; 
  return queue;
}

int push(Queue* queue, int elem) {
  Node* node = malloc(sizeof(Node));
  node->data = elem; 
  if(queue->end == NULL) {
    queue->end = node;
  } else {
    queue->end->next = node;
    queue->end = node;
  }

  if(queue->start == NULL) {
    queue->start = node;
  }

  queue->size++;
  return queue->end->data;
}

int pop(Queue* queue) {
  if(queue->start != NULL) {
    Node* elem = queue->start;
    queue->start = queue->start->next;
    queue->size--;
    free(elem);
    return elem->data;
  }
  printf("Queue is empty!\n");
  return -1;
}

int peek(Queue* queue) {
  if(queue->start != NULL) {
    return queue->start->data;
  }
  printf("Queue is empty!\n");
  return -1;
}

int main(void) {
  Queue* queue = constructor();
  printf("----- PUSHING ELEMENTS ------\n");
  push(queue, 1);
  push(queue, 2);
  push(queue, 7);
  push(queue, 8);
  push(queue, 9);
  print(queue);
  printf("----- POPPING ELEMENTS ------\n");
  pop(queue);
  print(queue);
  printf("----- PEEKING ELEMENTS ------\n");
  printf("%d\n", peek(queue));
  return 0;
}